pkg/pillar/dpcreconciler/linux.go

```diff
@@ -1406,23 +1444,92 @@ func (r *LinuxDpcReconciler) getIntendedWwanConfig(dpc types.DevicePortConfig,
+			PreferredRATs:     accessPoint.PreferredRATs,
+			ForbidRoaming:     accessPoint.ForbidRoaming,
+			Proxies:           port.Proxies,
+			Probe:             probeCfg,
+			LocationTracking:  locationTracking,
 		}
 		config.Networks = append(config.Networks, network)
 	}
 	return generic.Wwan{Config: config}
 }
 
+func (r *LinuxDpcReconciler) getWwanCredentials(ap *types.CellularAccessPoint) (
+	decBlock types.EncryptionBlock, err error) {
+	if !ap.EncryptedCredentials.IsCipher {
+		return decBlock, nil
+	}
+	decryptAvailable := r.SubControllerCert != nil && r.SubEdgeNodeCert != nil
+	if !decryptAvailable {
+		r.CipherMetrics.RecordFailure(r.Log, types.NotReady)
+		return decBlock, fmt.Errorf(
+			"missing certificates for decryption of cellular network credentials")
+	}
+	status, decBlock, err := cipher.GetCipherCredentials(
+		&cipher.DecryptCipherContext{
+			Log:               r.Log,
+			AgentName:         r.AgentName,
+			AgentMetrics:      r.CipherMetrics,
+			SubControllerCert: r.SubControllerCert,
+			SubEdgeNodeCert:   r.SubEdgeNodeCert,
+		},
+		ap.EncryptedCredentials)
+	if r.PubCipherBlockStatus != nil {
+		r.PubCipherBlockStatus.Publish(status.Key(), status)
+	}
+	if err != nil {
+		r.CipherMetrics.RecordFailure(r.Log, types.DecryptFailed)
+		return decBlock, fmt.Errorf(
+			"failed to decrypt cellular network credentials: %v", err)
+	}
+	return decBlock, nil
+}
+
+// Encrypt user password using AES-256-CBC with the key derived by the PBKDF2
+// method, taking kernel-generated /proc/sys/kernel/random/boot_id as the input.
+// Note that even though the config with the password is passed from NIM to the wwan
+// microservice using the *in-memory only* /run filesystem, we still encrypt
+// the password to avoid accidental exposure when the content of /run/wwan
+// is dumped as part of a customer issue report.
+func (r *LinuxDpcReconciler) encryptCellularPassword(portLL, password string) (
+	ciphertext string, err error) {
+	// Key includes port logical label so that even if user uses the same password
+	// for multiple APNs, the encrypted passwords are different for more secrecy.
+	key := portLL + "/" + password
+	ciphertext, alreadyEncrypted := r.encCellularPasswords[key]
+	if alreadyEncrypted {
+		// EncryptBytes uses random salt meaning that it will return different
+		// ciphertext on each run even for the same password.
+		// However, we do not want to change generated /run/wwan/config.json unless
+		// there is an actual config change (triggers many operations in the wwan
+		// microservice), so we remember already encrypted passwords.
+		return ciphertext, nil
+	}
+	bootID, err := os.ReadFile(bootIDFile)
+	if err != nil {
+		return "", err
+	}
+	passphrase := strings.TrimSpace(string(bootID))
+	openSSL := openssl.New()
+	encBytes, err := openSSL.EncryptBytes(
+		passphrase, []byte(password), openssl.PBKDF2SHA256)
+	if err != nil {
+		return "", err
+	}
+	ciphertext = string(encBytes)
+	r.encCellularPasswords[key] = ciphertext
```

Comment:
If folks rekey i.e., password argument gets updated you keep on adding new entry in this map. Should we be concerned about it growing and delete old entries for the same portLL?

Reply:
I have improved this to GC obsolete entries.